source/server/connection_handler_impl.cc

```diff
@@ -52,133 +56,142 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+                                                          per_address_details);
+          }
         }
       }
+    } else if (absl::holds_alternative<std::reference_wrapper<Network::InternalListener>>(
+                   per_address_details->typed_listener_)) {
+      internal_listener_map_by_address_.insert_or_assign(
+          per_address_details->address_->asStringView(), per_address_details);
     }
-  } else if (absl::holds_alternative<std::reference_wrapper<Network::InternalListener>>(
-                 details->typed_listener_)) {
-    internal_listener_map_by_address_.insert_or_assign(
-        config.listenSocketFactory().localAddress()->asStringView(), details);
   }
+  listener_map_by_tag_.emplace(config.listenerTag(), std::move(details));
 }
 
 void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   if (auto listener_iter = listener_map_by_tag_.find(listener_tag);
       listener_iter != listener_map_by_tag_.end()) {
     // listener_map_by_address_ may already update to the new listener. Compare it with the one
     // which find from listener_map_by_tag_, only delete it when it is same listener.
-    auto& address = listener_iter->second->address_;
-    auto address_view = address->asStringView();
-    if (tcp_listener_map_by_address_.contains(address_view) &&
-        tcp_listener_map_by_address_[address_view]->listener_tag_ ==
-            listener_iter->second->listener_tag_) {
-      tcp_listener_map_by_address_.erase(address_view);
-
-      // If the address is Ipv6 and isn't v6only, delete the corresponding Ipv4 item from the map.
-      if (address->type() == Network::Address::Type::Ip &&
-          address->ip()->version() == Network::Address::IpVersion::v6 &&
-          !address->ip()->ipv6()->v6only()) {
-        if (address->ip()->isAnyAddress()) {
-          auto ipv4_any_addr_iter = tcp_listener_map_by_address_.find(
-              Network::Address::Ipv4Instance(address->ip()->port()).asStringView());
-          // Since both "::" with ipv4_compat and "0.0.0.0" can be supported, ensure they are same
-          // listener by tag.
-          if (ipv4_any_addr_iter != tcp_listener_map_by_address_.end() &&
-              ipv4_any_addr_iter->second->listener_tag_ == listener_iter->second->listener_tag_) {
-            tcp_listener_map_by_address_.erase(ipv4_any_addr_iter);
-          }
-        } else {
-          auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
-          // Remove this check when runtime flag
-          // `envoy.reloadable_features.strict_check_on_ipv4_compat` deprecated.
-          if (v4_compatible_addr != nullptr) {
-            // both "::FFFF:<ipv4-addr>" with ipv4_compat and "<ipv4-addr>" isn't valid case,
-            // remove the v4 compatible addr item directly.
-            tcp_listener_map_by_address_.erase(v4_compatible_addr->asStringView());
+    for (auto& per_address_details : listener_iter->second->per_address_details_list_) {
```

Comment:
nit: can per_address_details be `const`?

Reply:
yes, it can be, let me update